io-engine/src/bdev/nexus/nexus_channel.rs

```diff
@@ -209,16 +211,66 @@ impl<'n> NexusChannel<'n> {
         }
     }
 
-    /// Disconnects a child device from the I/O path.
-    pub fn disconnect_device(&mut self, device_name: &str) {
+    /// Detaches a child device from this I/O channel, moving the device's
+    /// handles to the list of detached devices to disconnect later.
+    ///
+    /// The detached handles must be disconnected and dropped by a
+    /// `disconnect_detached_devices()` call.
+    pub(super) fn detach_device(&mut self, device_name: &str) {
         self.previous_reader = UnsafeCell::new(0);
 
-        self.readers
-            .retain(|c| c.get_device().device_name() != device_name);
-        self.writers
-            .retain(|c| c.get_device().device_name() != device_name);
+        if let Some(d) = self
+            .readers
+            .iter()
+            .position(|c| c.get_device().device_name() == device_name)
+        {
+            let t = self.readers.remove(d);
+            self.detached.push(t);
+        }
+
+        if let Some(d) = self
+            .writers
+            .iter()
+            .position(|c| c.get_device().device_name() == device_name)
+        {
+            let t = self.writers.remove(d);
+            self.detached.push(t);
+        }
+
+        debug!("{self:?}: device '{device_name}' detached");
+    }
+
+    /// Disconnects previously detached device handles by dropping them.
+    /// Devices to drop are filtered by the given predicate: true to drop
+    /// a device, false to keep it.
+    pub(super) fn disconnect_detached_devices<F>(&mut self, mut drop_pred: F)
+    where
+        F: FnMut(&dyn BlockDeviceHandle) -> bool,
+    {
+        let cnt = self
```

Comment:
What's the reason we pass a closure here(always returning `true`), then filter to count? and not simply check vector's length? In other words, when do we want to disconnect selectively from `detached` list?

Reply:
For debugging purposed only. I want to print a nice log messages here, with number of device to drop, devices remaining, etc.

Reply:
Perhaps we should do minimal required here because this path is latency sensitive i.e. we'd want the device retire to finish as soon as possible and not  hold the pause longer than needed.

Reply:
Ah, I just found a better way to report and avoid counting, I will print "before len" - "after len".